File: src/helpers/Executor.cpp
//MIT License

//Copyright (c) 2020 bexoft GmbH (mail@bexoft.de)

//Permission is hereby granted, free of charge, to any person obtaining a copy
//of this software and associated documentation files (the "Software"), to deal
//in the Software without restriction, including without limitation the rights
//to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//copies of the Software, and to permit persons to whom the Software is
//furnished to do so, subject to the following conditions:

//The above copyright notice and this permission notice shall be included in all
//copies or substantial portions of the Software.

//THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//SOFTWARE.

#include "finalmq/helpers/Executor.h"

#include <iostream>

#include <assert.h>

namespace finalmq
{
void ExecutorBase::registerActionNotification(std::function<void()> func)
{
    m_funcNotify = func;
}

void ExecutorBase::run()
{
    while (!m_terminate.load())
    {
        bool wasAvailable = runAvailableActionBatch([this]() {
            return m_terminate.load();
        });
        if (!wasAvailable && !m_terminate.load())
        {
            m_newActions.wait();
        }
    }
    // release possible other threads
    m_newActions = true;
}

void ExecutorBase::terminate()
{
    m_terminate = true;
    m_newActions = true;
}

bool ExecutorBase::isTerminating() const
{
    return m_terminate;
}

////////////////////////////////////////////////////////

bool Executor::runAvailableActions(const FuncIsAbort& funcIsAbort)
{
    std::list<ActionEntry> actions;
    std::unique_lock<std::mutex> lock(m_mutex);
    actions = std::move(m_actions);
    m_actions.clear();
    m_zeroIdCounter = 0;
    m_storedIds.clear();
    m_runningIds.clear();
    lock.unlock();

    if (!actions.empty())
    {
        for (auto it1 = actions.begin(); it1 != actions.end(); ++it1)
        {
            ActionEntry& entry = *it1;
            for (auto it2 = entry.funcs.begin(); it2 != entry.funcs.end(); ++it2)
            {
                std::unique_ptr<std::function<void()>>& func = *it2;
                assert(func && *func);
                if (!funcIsAbort || !funcIsAbort())
                {
                    (*func)();
                }
                else
                {
                    return true;
                }
            }
        }
        return true;
    }
    else
    {
        return false;
    }
}

bool Executor::areRunnableActionsAvailable() const
{
    if (m_runningIds.size() == m_storedIds.size() && (m_zeroIdCounter == 0))
    {
        return false;
    }
    return true;
}

bool Executor::runAvailableActionBatch(const FuncIsAbort& funcIsAbort)
{
    bool wasAvailable = false;
    bool stillActions = false;
    std::unique_lock<std::mutex> lock(m_mutex);
    if (!areRunnableActionsAvailable())
    {
        return false;
    }
    std::deque<std::unique_ptr<std::function<void()>>> funcs;
    std::int64_t instanceId = -1;
    for (auto it = m_actions.begin(); it != m_actions.end(); ++it)
    {
        ActionEntry& entry = *it;
        if (entry.instanceId == 0 || m_runningIds.find(entry.instanceId) == m_runningIds.end())
        {
            instanceId = entry.instanceId;

            bool erased = false;
            if (instanceId != 0)
            {
                m_runningIds.insert(instanceId);
                funcs = std::move(entry.funcs);
                it = m_actions.erase(it);
                erased = true;
            }
            else
            {
                --m_zeroIdCounter;
                funcs.push_back(std::move(entry.funcs.front()));
                if (entry.funcs.size() == 1)
                {
                    it = m_actions.erase(it);
                    erased = true;
                }
                else
                {
                    entry.funcs.pop_front();
                }
            }
            if (erased && it != m_actions.end())
            {
                auto itPrev = it;
                if (itPrev != m_actions.begin())
                {
                    --itPrev;
                    if (itPrev->instanceId == it->instanceId)
                    {
                        itPrev->funcs.insert(itPrev->funcs.end(), std::make_move_iterator(it->funcs.begin()), std::make_move_iterator(it->funcs.end()));
                        m_actions.erase(it);
                    }
                }
            }
            wasAvailable = true;
            stillActions = areRunnableActionsAvailable();
            break;
        }
    }
    lock.unlock();

    // trigger next possible thread
    if (stillActions)
    {
        m_newActions = true;
    }

    for (auto it = funcs.begin(); it != funcs.end(); ++it)
    {
        assert(*it && **it);
        if (!funcIsAbort || !funcIsAbort())
        {
            (**it)();
        }
        else
        {
            break;
        }
    }

    if (instanceId > 0)
    {
        lock.lock();
        m_runningIds.erase(instanceId);
        auto it = m_storedIds.find(instanceId);
        assert(it != m_storedIds.end());
        auto& counter = it->second;
        assert(counter > 0);
        assert(counter >= static_cast<std::int32_t>(funcs.size()));
        counter -= static_cast<std::int32_t>(funcs.size());
        if (counter == 0)
        {
            m_storedIds.erase(it);
        }
        lock.unlock();
    }
    return wasAvailable;
}

void Executor::addAction(std::function<void()> func, std::int64_t instanceId)
{
    bool notify = false;
    std::unique_lock<std::mutex> lock(m_mutex);
    if (instanceId != 0)
    {
        auto& count = m_storedIds[instanceId];
        notify = (count == 0);
        ++count;
    }
    else
    {
        notify = (m_zeroIdCounter == 0);
        ++m_zeroIdCounter;
    }
    if (!m_actions.empty() && m_actions.back().instanceId == instanceId)
    {
        m_actions.back().funcs.push_back(std::make_unique<std::function<void()>>(std::move(func)));
    }
    else
    {
        m_actions.emplace_back(instanceId, std::make_unique<std::function<void()>>(std::move(func)));
    }
    lock.unlock();
    if (notify)
    {
        m_newActions = true;
        if (m_funcNotify)
        {
            m_funcNotify();
        }
    }
}

//////////////////////////////////////////////////

bool ExecutorIgnoreOrderOfInstance::runAvailableActions(const FuncIsAbort& funcIsAbort)
{
    std::deque<std::function<void()>> actions;
    std::unique_lock<std::mutex> lock(m_mutex);
    actions = std::move(m_actions);
    m_actions.clear();
    lock.unlock();
    if (!actions.empty())
    {
        for (size_t i = 0; i < actions.size(); ++i)
        {
            if (!funcIsAbort || !funcIsAbort())
            {
                actions[i]();
            }
            else
            {
                break;
            }
        }
        return true;
    }
    else
    {
        return false;
    }
}

bool ExecutorIgnoreOrderOfInstance::runAvailableActionBatch(const FuncIsAbort& funcIsAbort)
{
    bool wasAvailable = false;
    bool stillActions = false;
    std::unique_lock<std::mutex> lock(m_mutex);
    std::function<void()> action;
    if (!m_actions.empty())
    {
        action = std::move(m_actions.front());
        m_actions.pop_front();
        wasAvailable = true;
        stillActions = (!m_actions.empty());
    }
    lock.unlock();
    if (stillActions)
    {
        m_newActions = true;
    }
    if (action)
    {
        if (!funcIsAbort || !funcIsAbort())
        {
            action();
        }
    }

    return wasAvailable;
}

void ExecutorIgnoreOrderOfInstance::addAction(std::function<void()> func, std::int64_t /*instanceId*/)
{
    std::unique_lock<std::mutex> lock(m_mutex);
    bool notify = m_actions.empty();
    m_actions.push_back(std::move(func));
    lock.unlock();
    if (notify)
    {
        m_newActions = true;
        if (m_funcNotify)
        {
            m_funcNotify();
        }
    }
}

//////////////////////////////////////////////////

ExecutorWorkerBase::ExecutorWorkerBase(const std::shared_ptr<IExecutor>& executor, int numberOfWorkerThreads)
    : m_executor(executor)
{
    for (int i = 0; i < numberOfWorkerThreads; ++i)
    {
        m_threads.emplace_back(std::thread([this]() {
            m_executor->run();
        }));
    }
}

ExecutorWorkerBase::~ExecutorWorkerBase()
{
    m_executor->terminate();
    join();
}

IExecutorPtr ExecutorWorkerBase::getExecutor() const
{
    return m_executor;
}

void ExecutorWorkerBase::addAction(std::function<void()> func, std::int64_t instanceId)
{
    m_executor->addAction(std::move(func), instanceId);
}

void ExecutorWorkerBase::terminate()
{
    m_executor->terminate();
}

bool ExecutorWorkerBase::isTerminating() const
{
    return m_executor->isTerminating();
}

void ExecutorWorkerBase::join()
{
    for (size_t i = 0; i < m_threads.size(); ++i)
    {
        m_threads[i].join();
    }
}

/////////////////////////////////////////////////////

void GlobalExecutorWorker::setInstance(std::unique_ptr<IExecutorWorker>&& instanceUniquePtr)
{
    getStaticUniquePtrRef() = std::move(instanceUniquePtr);
    getStaticInstanceRef().store(getStaticUniquePtrRef().get(), std::memory_order_release);
}

IExecutorWorker* GlobalExecutorWorker::createInstance()
{
    static std::mutex mutex;
    std::unique_lock<std::mutex> lock(mutex);
    IExecutorWorker* inst = getStaticInstanceRef().load(std::memory_order_relaxed);
    if (!inst)
    {
        setInstance(std::make_unique<ExecutorWorker<Executor>>());
        inst = getStaticInstanceRef().load(std::memory_order_relaxed);
    }
    return inst;
}

std::atomic<IExecutorWorker*>& GlobalExecutorWorker::getStaticInstanceRef()
{
    static std::atomic<IExecutorWorker*> instance;
    return instance;
}

std::unique_ptr<IExecutorWorker>& GlobalExecutorWorker::getStaticUniquePtrRef()
{
    static std::unique_ptr<IExecutorWorker> instanceUniquePtr;
    return instanceUniquePtr;
}

} // namespace finalmq

File: csharp/src/helper/Executor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Threading;
using System.Diagnostics;

namespace finalmq
{
    abstract class ExecutorBase : IExecutor
    {
        public void RegisterActionNotification(DelegateNotification func)
        {
            m_funcNotify = func;
        }

        public void Run()
        {
            while (Interlocked.Read(ref m_terminate) == 0)
            {
                bool wasAvailable = RunAvailableActionBatch(() => {
                    return (Interlocked.Read(ref m_terminate) != 0);
                });
                if (!wasAvailable && (Interlocked.Read(ref m_terminate) == 0))
                {
                    m_newActions.Wait();
                }
            }
            // release possible other threads
            m_newActions.Set();
        }

        public void Terminate()
        {
            Interlocked.Exchange(ref m_terminate, 1);
            m_newActions.Set();
        }

        public bool IsTerminating()
        {
            return (Interlocked.Read(ref m_terminate) != 0);
        }

        public abstract bool RunAvailableActions(DelegateIsAbort? funcIsAbort = null);
        public abstract bool RunAvailableActionBatch(DelegateIsAbort? funcIsAbort = null);
        public abstract void AddAction(DelegateAction func, Int64 instanceId);

        long m_terminate = 0;    // atomic
        protected CondVar m_newActions = new CondVar();
        protected DelegateNotification? m_funcNotify = null;
        protected Object m_mutex = new Object();
    }

    class Executor : ExecutorBase
    {
        public override bool RunAvailableActions(DelegateIsAbort? funcIsAbort = null)
        {
            IList<ActionEntry> actions;
            lock (m_mutex)
            {
                actions = m_actions;
                m_actions = new List<ActionEntry>();
                m_zeroIdCounter = 0;
                m_storedIds.Clear();
                m_runningIds.Clear();
            }

            if (actions.Count != 0)
            {
                foreach (var entry in actions)
                {
                    foreach (var func in entry.funcs)
                    {
                        Debug.Assert(func != null);
                        if (funcIsAbort == null || !funcIsAbort())
                        {
                            func();
                        }
                        else
                        {
                            return true;
                        }
                    }
                }
                return true;
            }
            else
            {
                return false;
            }
        }

        public override bool RunAvailableActionBatch(DelegateIsAbort? funcIsAbort = null)
        {
            bool wasAvailable = false;
            bool stillActions = false;
            IList<DelegateAction> funcs = new List<DelegateAction>();
            long instanceId = -1;
            lock (m_mutex)
            {
                if (!AreRunnableActionsAvailable())
                {
                    return false;
                }
                for (int i = 0; i < m_actions.Count; ++i)
                {
                    ActionEntry entry = m_actions[i];
                    if (entry.instanceId == 0 || !m_runningIds.Contains(entry.instanceId))
                    {
                        instanceId = entry.instanceId;

                        bool erased = false;
                        if (instanceId != 0)
                        {
                            m_runningIds.Add(instanceId);
                            funcs = entry.funcs.ToList();
                            m_actions.RemoveAt(i);
                            erased = true;
                        }
                        else
                        {
                            --m_zeroIdCounter;
                            funcs.Add(entry.funcs[0]);
                            if (entry.funcs.Count == 1)
                            {
                                m_actions.RemoveAt(i);
                                erased = true;
                            }
                            else
                            {
                                entry.funcs.RemoveAt(0);
                            }
                        }
                        if (erased && i < m_actions.Count)
                        {
                            int iPrev = i - 1;
                            if (iPrev >= 0)
                            {
                                if (m_actions[iPrev].instanceId == m_actions[i].instanceId)
                                {
                                    m_actions[iPrev].funcs.AddRange(m_actions[i].funcs);
                                    m_actions.RemoveAt(i);
                                }
                            }
                        }
                        wasAvailable = true;
                        stillActions = AreRunnableActionsAvailable();
                        break;
                    }
                }
            }

            // trigger next possible thread
            if (stillActions)
            {
                m_newActions.Set();
            }

            foreach (var func in funcs)
            {
                if (funcIsAbort == null || !funcIsAbort())
                {
                    func();
                }
                else
                {
                    break;
                }
            }

            if (instanceId > 0)
            {
                lock (m_mutex)
                {
                    m_runningIds.Remove(instanceId);
                    int counter;
                    bool found = m_storedIds.TryGetValue(instanceId, out counter);
                    Debug.Assert(found);
                    Debug.Assert(counter > 0);
                    Debug.Assert(counter >= funcs.Count);
                    counter -= funcs.Count;
                    if (counter == 0)
                    {
                        m_storedIds.Remove(instanceId);
                    }
                }
            }
            return wasAvailable;
        }

        public override void AddAction(DelegateAction func, long instanceId = 0)
        {
            bool notify = false;
            lock (m_mutex)
            {
                if (instanceId != 0)
                {
                    System.Int32 count = m_storedIds[instanceId];
                    notify = (count == 0);
                    ++count;
                }
                else
                {
                    notify = (m_zeroIdCounter == 0);
                    ++m_zeroIdCounter;
                }
                if ((m_actions.Count != 0) && m_actions.Last().instanceId == instanceId)
                {
                    m_actions.Last().funcs.Add(func);
                }
                else
                {
                    m_actions.Add(new ActionEntry(instanceId, func));
                }
            }
            if (notify)
            {
                m_newActions.Set();
                if (m_funcNotify != null)
                {
                    m_funcNotify();
                }
            }
        }

        bool AreRunnableActionsAvailable()
        {
            if (m_runningIds.Count() == m_storedIds.Count() && (m_zeroIdCounter == 0))
            {
                return false;
            }
            return true;
        }

        class ActionEntry
        {
            public ActionEntry(long i, DelegateAction f)
            {
                instanceId = i;
                funcs.Add(f);
            }
            public long instanceId;
            public List<DelegateAction> funcs = new List<DelegateAction>();
        };

        IList<ActionEntry> m_actions = new List<ActionEntry>();

        IDictionary<long, int> m_storedIds = new Dictionary<long, int>();
        ISet<long> m_runningIds = new HashSet<long>();
        int m_zeroIdCounter = 0;
    }

    class ExecutorIgnoreOrderOfInstance : ExecutorBase
    {
        public override bool RunAvailableActions(DelegateIsAbort? funcIsAbort = null)
        {
            IList<DelegateAction> actions;
            lock (m_mutex)
            {
                actions = m_actions;
                m_actions = new List<DelegateAction>();
            }
            if (actions.Count != 0)
            {
                foreach (var action in m_actions)
                {
                    if (funcIsAbort == null || !funcIsAbort())
                    {
                        action();
                    }
                    else
                    {
                        break;
                    }
                }
                return true;
            }
            else
            {
                return false;
            }
        }

        public override bool RunAvailableActionBatch(DelegateIsAbort? funcIsAbort = null)
        {
            bool wasAvailable = false;
            bool stillActions = false;
            DelegateAction? action = null;
            lock (m_mutex)
            {
                if (m_actions.Count != 0)
                {
                    action = m_actions[0];
                    m_actions.RemoveAt(0);
                    wasAvailable = true;
                    stillActions = (m_actions.Count != 0);
                }
            }
            if (stillActions)
            {
                m_newActions.Set();
            }
            if (action != null)
            {
                if (funcIsAbort == null || !funcIsAbort())
                {
                    action();
                }
            }

            return wasAvailable;
        }

        public override void AddAction(DelegateAction func, long instanceId = 0)
        {
            bool notify = false;
            lock (m_mutex)
            {
                notify = (m_actions.Count == 0);
                m_actions.Add(func);
            }
            if (notify)
            {
                m_newActions.Set();
                if (m_funcNotify != null)
                {
                    m_funcNotify();
                }
            }
        }

        private IList<DelegateAction> m_actions = new List<DelegateAction>();
    }

    class ExecutorWorkerBase : IExecutorWorker
    {

        public ExecutorWorkerBase(IExecutor executor, int numberOfWorkerThreads = 4)
        {
            m_executor = executor;
            for (int i = 0; i < numberOfWorkerThreads; ++i)
            {
                Thread thread = new Thread(x =>
                {
                    m_executor.Run();
                });
                m_threads.Add(thread);
                thread.Start();
            }
        }

        ~ExecutorWorkerBase()
        {
            m_executor.Terminate();
            Join();
        }

        public IExecutor Executor
        {
            get
            {
                return m_executor;
            }
        }
        public void AddAction(DelegateAction func, Int64 instanceId = 0)
        {
            m_executor.AddAction(func, instanceId);
        }
        public void Terminate()
        {
            m_executor.Terminate();
        }
        public bool IsTerminating()
        {
            return m_executor.IsTerminating();
        }
        public void Join()
        {
            foreach (var t in m_threads)
            {
                t.Join();
            }
        }

        private IExecutor       m_executor;
        private IList<Thread>   m_threads = new List<Thread>();
    }

    class ExecutorWorker<T> : ExecutorWorkerBase where T : IExecutor, new()
    {
        public ExecutorWorker(int numberOfWorkerThreads = 4)
            : base(new T(), numberOfWorkerThreads)
        {
        }
    };

    class GlobalExecutorWorker
    {
        public static IExecutorWorker Instance
        {
            get => m_instance;
            set
            {
                m_instance = value;
            }
        }

        private GlobalExecutorWorker()
        {
        }

        private static IExecutorWorker m_instance = new ExecutorWorker<Executor>();
    }

}




